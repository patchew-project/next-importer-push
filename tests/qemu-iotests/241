#!/usr/bin/env bash
# group: rw quick
#
# Test qemu-nbd vs. unaligned images
#
# Copyright (C) 2018-2021 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

seq="$(basename $0)"
echo "QA output created by $seq"

status=1 # failure is the default!

_cleanup()
{
    _cleanup_test_img
    rm -f "$TEST_DIR/server.log" "$TEST_IMG_FILE.mid" "$TEST_IMG_FILE.qcow2"
    nbd_server_stop
}
trap "_cleanup; exit \$status" 0 1 2 3 15

# get standard environment, filters and checks
. ./common.rc
. ./common.filter
. ./common.nbd

_supported_fmt raw # although the test also requires use of qcow2
_supported_proto nbd
_supported_os Linux
_require_command QEMU_NBD

# can't use _make_test_img, because qemu-img rounds image size up,
# and because we want to use Unix socket rather than TCP port. Likewise,
# we have to redirect TEST_IMG to our server.
# This tests that we can deal with the hole at the end of an unaligned
# raw file (either because the server doesn't advertise alignment too
# large, or because the client ignores the server's noncompliance - even
# though we can't actually wire iotests into checking trace messages).
printf %01000d 0 > "$TEST_IMG_FILE"
TEST_IMG="nbd:unix:$nbd_unix_socket"

echo
echo "=== Exporting unaligned raw image, natural alignment ==="
echo

nbd_server_start_unix_socket -f $IMGFMT "$TEST_IMG_FILE"

$QEMU_NBD_PROG --list -k $nbd_unix_socket | grep '\(size\|min\)'
$QEMU_IMG map -f raw --output=json "$TEST_IMG" | _filter_qemu_img_map
$QEMU_IO -f raw -c map "$TEST_IMG"
nbd_server_stop

echo
echo "=== Exporting unaligned raw image, forced server sector alignment ==="
echo

# Intentionally omit '-f' to force image probing, which in turn forces
# sector alignment, here at the server.
nbd_server_start_unix_socket "$TEST_IMG_FILE" 2> "$TEST_DIR/server.log"

$QEMU_NBD_PROG --list -k $nbd_unix_socket | grep '\(size\|min\)'
$QEMU_IMG map -f raw --output=json "$TEST_IMG" | _filter_qemu_img_map
$QEMU_IO -f raw -c map "$TEST_IMG"
nbd_server_stop
cat "$TEST_DIR/server.log" | _filter_testdir

echo
echo "=== Exporting unaligned raw image, forced client sector alignment ==="
echo

# Now force sector alignment at the client.
nbd_server_start_unix_socket -f $IMGFMT "$TEST_IMG_FILE"

$QEMU_NBD_PROG --list -k $nbd_unix_socket | grep '\(size\|min\)'
$QEMU_IMG map --output=json "$TEST_IMG" | _filter_qemu_img_map
$QEMU_IO -c map "$TEST_IMG"
nbd_server_stop

echo
echo "=== Encrypted qcow2 file backed by unaligned raw image ==="
echo

# Enabling encryption in qcow2 forces 512-alignment
SECRET=secret,id=sec0,data=12345
$QEMU_IMG create -f qcow2 -b "$TEST_IMG_FILE" -F raw --object "$SECRET" \
  -o encrypt.format=luks,encrypt.key-secret=sec0,encrypt.iter-time=10 \
  "$TEST_IMG_FILE.qcow2" 2k | _filter_img_create
nbd_server_start_unix_socket --object "$SECRET" --image-opts \
  driver=qcow2,file.filename="$TEST_IMG_FILE.qcow2",encrypt.key-secret=sec0

$QEMU_NBD_PROG --list -k $nbd_unix_socket | grep '\(size\|min\)'
$QEMU_IMG map --output=json "$TEST_IMG" | _filter_qemu_img_map
$QEMU_IO -c map "$TEST_IMG"
nbd_server_stop

echo
echo "=== Use blkdebug for larger alignment than backing layer ==="
echo

$QEMU_IMG create -f qcow2 -o cluster_size=1024 \
  "$TEST_IMG_FILE" 2k | _filter_img_create
$QEMU_IMG create -f qcow2 -b "$TEST_IMG_FILE" -F qcow2 -o cluster_size=512 \
  "$TEST_IMG_FILE.mid" | _filter_img_create
$QEMU_IMG bitmap --add -g 512 --enable "$TEST_IMG_FILE.mid" b0
$QEMU_IO -f qcow2 -c 'w 512 512' "$TEST_IMG_FILE.mid" | _filter_qemu_io
$QEMU_IMG create -f qcow2 -b "$TEST_IMG_FILE.mid" -F qcow2 \
  "$TEST_IMG_FILE.qcow2" 4k | _filter_img_create
FILE=image.file.filename="$TEST_IMG_FILE.qcow2"
nbd_server_start_unix_socket -B b0 -A --image-opts \
  driver=blkdebug,align=4096,image.driver=qcow2,image.file.driver=file,"$FILE"

TEST_IMG="driver=nbd,server.type=unix,server.path=$nbd_unix_socket"
$QEMU_NBD_PROG --list -k $nbd_unix_socket | grep '\(size\|min\)'
# Reports a single 4k block of "data":false, meaning dirty.  Reporting
# "data":true would be wrong (the entire region is not clean).
$QEMU_IMG map --output=json --image-opts \
	  "$TEST_IMG",x-dirty-bitmap=qemu:dirty-bitmap:b0 | _filter_qemu_img_map
# Reports a single 4k block of "zero":true,"data":true, meaning allocated
# from the backing chain.  Reporting "zero":false,"data":false would be wrong
# (nothing is allocated in the active layer), and as would reporting
# "zero":false,"data":true (the entire region is not unallocated).
$QEMU_IMG map --output=json --image-opts \
  "$TEST_IMG",x-dirty-bitmap=qemu:allocation-depth | _filter_qemu_img_map
nbd_server_stop

# Not tested here: we also want to ensure that qemu as NBD client does
# not access beyond the end of a server's advertised unaligned size:
#  nbdkit -U - memory size=513 --run 'qemu-io -f raw -c "r 512 512" $nbd'
# However, since qemu as server always rounds up to a sector alignment,
# we would have to use nbdkit to demonstrate this scenario (see
# commit 9cf638508c for more information).

# success, all done
echo '*** done'
rm -f $seq.full
status=0
