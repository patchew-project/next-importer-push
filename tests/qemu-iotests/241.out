QA output created by 241

=== Exporting unaligned raw image, natural alignment ===

  size:  1024
  min block: 1
[{ "start": 0, "length": 1000, "depth": 0, "zero": false, "data": true, "offset": OFFSET},
{ "start": 1000, "length": 24, "depth": 0, "zero": true, "data": true, "offset": OFFSET}]
1 KiB (0x400) bytes     allocated at offset 0 bytes (0x0)

=== Exporting unaligned raw image, forced server sector alignment ===

  size:  1024
  min block: 512
[{ "start": 0, "length": 1024, "depth": 0, "zero": false, "data": true, "offset": OFFSET}]
1 KiB (0x400) bytes     allocated at offset 0 bytes (0x0)
WARNING: Image format was not specified for 'TEST_DIR/t.raw' and probing guessed raw.
         Automatically detecting the format is dangerous for raw images, write operations on block 0 will be restricted.
         Specify the 'raw' format explicitly to remove the restrictions.

=== Exporting unaligned raw image, forced client sector alignment ===

  size:  1024
  min block: 1
[{ "start": 0, "length": 1000, "depth": 0, "zero": false, "data": true, "offset": OFFSET},
{ "start": 1000, "length": 24, "depth": 0, "zero": true, "data": true, "offset": OFFSET}]
1 KiB (0x400) bytes     allocated at offset 0 bytes (0x0)

=== Encrypted qcow2 file backed by unaligned raw image ===

Formatting 'TEST_DIR/t.IMGFMT.qcow2', fmt=qcow2 size=2048 backing_file=TEST_DIR/t.IMGFMT backing_fmt=IMGFMT
  size:  2048
  min block: 512
[{ "start": 0, "length": 1024, "depth": 0, "zero": false, "data": true, "offset": OFFSET},
{ "start": 1024, "length": 1024, "depth": 0, "zero": true, "data": true, "offset": OFFSET}]
2 KiB (0x800) bytes     allocated at offset 0 bytes (0x0)

=== Use blkdebug for larger alignment than backing layer ===

Formatting 'TEST_DIR/t.IMGFMT', fmt=qcow2 size=2048
Formatting 'TEST_DIR/t.IMGFMT.mid', fmt=qcow2 size=2048 backing_file=TEST_DIR/t.IMGFMT backing_fmt=qcow2
wrote 512/512 bytes at offset 512
512 bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
Formatting 'TEST_DIR/t.IMGFMT.qcow2', fmt=qcow2 size=4096 backing_file=TEST_DIR/t.IMGFMT.mid backing_fmt=qcow2
  size:  4096
  min block: 4096
[{ "start": 0, "length": 4096, "depth": 0, "zero": false, "data": true, "offset": OFFSET}]
[{ "start": 0, "length": 4096, "depth": 0, "zero": false, "data": true, "offset": OFFSET}]
*** done
